src/compiler/compile/render_dom/wrappers/Element/index.ts

```diff
@@ -1119,19 +1141,39 @@ export default class ElementWrapper extends Wrapper {
 
 			block.chunks.hydrate.push(updater);
 
-			const dependencies = expression.dynamic_dependencies();
+			// Assume that style has changed through the spread attribute
 			if (has_spread) {
 				block.chunks.update.push(updater);
-			} else if (dependencies.length > 0) {
-				if (should_cache) {
-					block.chunks.update.push(b`
-							if (${block.renderer.dirty(dependencies)} && (${cached_snippet} !== (${cached_snippet} = ${snippet}))) {
+			} else {
+				const deps = [
+					...expression.dynamic_dependencies().filter(dep => !this.dynamic_style_dependencies.has(dep)),
+					...this.dynamic_style_dependencies
+				];
+
+				if (deps.length > 0) {
+					if (deps.length === this.dynamic_style_dependencies.size) {
+						maybe_create_style_changed_var();
+						block.chunks.update.push(b`
+							if (${style_changed_var}) {
 								${updater}
 							}
-					`);
-				} else {
+						`);
+						return;
+					}
+
+					let condition =  block.renderer.dirty(deps);
+
+					if (should_cache) {
+						condition = x`${condition} && ${cached_snippet} !== (${cached_snippet} = ${snippet})`;
+
+						maybe_create_style_changed_var();
+						if (style_changed_var) {
+							condition = x`${style_changed_var} || ${condition}`;
+						}
+					}
```

Comment:
I think we can reuse the same code for adding into the `block.chunks.update` by combining the logic into the `condition`:

```suggestion
				  let condition = block.renderer.dirty(deps);
				  
				  if (should_cache) {
						condition = x`${condition} && ${cached_snippet} !== (${cached_snippet} = ${snippet})`;
				  }
				  
				  // over here we can use this set to determine whether the style directive is depending on any style attribute's dependencies
				  if (dynamic_style_dependencies.size > 0) {
				    maybe_create_style_changed_var();
				    condition = x`${style_changed_var} || ${condition}`;
				  }
```